packages/build/src/utils/install.js

```diff
@@ -1,33 +1,54 @@
 const execa = require('execa')
+const pathExists = require('path-exists')
 
 const { addErrorInfo } = require('../error/info')
 
 // Install Node.js dependencies in a specific directory
-const installDependencies = async function({ packageRoot }) {
-  try {
-    await execa.command('npm install --no-progress --no-audit --no-fund', {
-      cwd: packageRoot,
-      all: true,
-    })
-  } catch (error) {
-    error.message = `Error while installing dependencies in ${packageRoot}\n${error.all}`
-    addErrorInfo(error, { type: 'dependencies' })
-    throw error
-  }
+const installDependencies = function({ packageRoot }) {
+  return runCommand({ packageRoot, type: 'install' })
 }
 
 // Add new Node.js dependencies
-const addDependencies = async function({ packageRoot, packages }) {
+const addDependencies = function({ packageRoot, packages }) {
+  return runCommand({ packageRoot, packages, type: 'add' })
+}
+
+const runCommand = async function({ packageRoot, packages, type }) {
   try {
-    await execa.command(`npm install --no-progress --no-audit --no-fund ${packages.join(' ')}`, {
-      cwd: packageRoot,
-      all: true,
-    })
+    const command = await getCommand(packageRoot, type)
+    const packagesList = packages === undefined ? '' : ` ${packages.join(' ')}`
+    await execa.command(`${command}${packagesList}`, { cwd: packageRoot, all: true })
   } catch (error) {
-    const errorA = new Error(`Error while installing dependencies in ${packageRoot}\n${error.all}`)
+    const message = error.code === 'ENOENT' ? `${error.path} is not installed` : error.all
+    const errorA = new Error(`Error while installing dependencies in ${packageRoot}\n${message}`)
     addErrorInfo(errorA, { type: 'dependencies' })
     throw errorA
   }
 }
 
+// Retrieve the shell command to install or add dependencies
+const getCommand = async function(packageRoot, type) {
+  const manager = await getManager(packageRoot)
+  return COMMANDS[manager][type]
+}
+
+const getManager = async function(packageRoot) {
+  if (await pathExists(`${packageRoot}/yarn.lock`)) {
+    return 'yarn'
+  }
+
+  return 'npm'
+}
+
+const COMMANDS = {
+  npm: {
+    install: 'npm install --no-progress --no-audit --no-fund',
+    add: 'npm install --no-progress --no-audit --no-fund',
+  },
+  yarn: {
+    install: 'yarn install --no-progress --non-interactive',
```

Comment:
Might be out of scope for this PR, but shouldn't we use`--frozen-lockfile` for `yarn` and `npm ci` (in case there is a lock file)?

Without `--frozen-lockfile` or `npm ci` lock files are updated to match `package.json` which creates non-reproducible builds.

Reply:
Yes you're right, we should get the lock files update question resolved in this PR.

I think the situation is different depending on where the install takes place.

In CI, if a lock file exists, we should use it but not update it. Otherwise we should not create one.

In local builds (e.g. Netlify CLI), if a lock file exists, we should use it and potentially update it. Otherwise we should create one. The idea is that when developers run `netlify build`, we are doing package management for them automatically, including semver package updates. This does mean we are modifying/adding source files, but developers would have to do it manually otherwise anyway (by going inside each local plugin and/or Netlify Functions and running `npm` or `yarn`). We have also updated several plugins `README` to remove the manual `npm install` step, so that seems to be the direction we are going towards product-wise.

If we follow the principle above, we've got 8 different commands depending on whether:
  - npm or yarn is used
  - a new package is installed, or all packages are being installed/updated
  - we are in CI or local

We will also need to run different commands depending on whether a lock file exists or not. For example, `npm ci` fails if there is no `package-lock.json`. However we don't want to require a lock file.

Also: 
  - `npm ci` does not work for installing new packages.
  - `npm --no-package-lock` prevents updating `package-lock.json`. This would be good in CI. However this flag also prevents using `package-lock.json`, so should only be used when there is no `package-lock.json`.

This would give the following set of commands:
 - CI:
    - npm:
       - all packages, if package-lock.json: `npm ci --no-progress --no-audit --no-fund`
       - all packages, if no package-lock.json: `npm install --no-progress --no-audit --no-fund --no-package-lock`
       - new packages, if package-lock.json: `npm install --no-progress --no-audit --no-fund PACKAGE...`
       - new packages, if no package-lock.json: `npm install --no-progress --no-audit --no-fund --no-package-lock PACKAGE...`
    - yarn:
       - all packages, if yarn.lock: `yarn install --no-progress --non-interactive --frozen-lockfile`
       - all packages, if no yarn.lock: `yarn install --no-progress --non-interactive --pure-lockfile`
       - new packages: `yarn add --no-progress --non-interactive --pure-lockfile PACKAGE...`
 - Local:
    - npm:
       - all packages: `npm install --no-progress --no-audit --no-fund`
       - new packages: `npm install --no-progress --no-audit --no-fund PACKAGE...`
    - yarn:
       - all packages: `yarn install --no-progress --non-interactive`
       - new packages: `yarn add --no-progress --non-interactive PACKAGE...`

What do you think?

Reply:
I'm good with that.

FYI, `yarn install --frozen-lockfile` doesn't seem to fail when there is no lock file, but I think we can keep using `yarn install --pure-lockfile` for clarity

Reply:
Good point: just updated my message :)

I also updated this: I actually think we should create lock files in CI if there are none. This is because lock files are sometimes used for caching purpose (using their hash/checksum). I.e. `yarn --pure-lockfile` and `npm --no-package-lock` should actually not be used when there is no lock file.

Let me know if all still looks good.

Reply:
I'm good with that, it is a bit annoying how `npm ci` and `yarn` behave differently for the same functionality.

Reply:
:100:

Reply:
Note: we detect Yarn by checking for a `yarn.lock`, i.e. when Yarn is used, there is always be a lock file.

Reply:
Other note: `npm ci` was introduced by `npm@5.7.0` which was shipped in `node@8.12.0` and `node@10.3.0`. We still support older versions of Node (such as `8.3.0`) so we need to:
  - run `npm --version`
  - if `npm` is older than `5.7.0`, use `npm ci` instead of `npm install`